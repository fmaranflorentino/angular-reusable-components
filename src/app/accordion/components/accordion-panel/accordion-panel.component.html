<div
  class="accordion_panel__header__container"
  (click)="handlePanelInteraction()"
>
  <app-accordion-panel-header [expanded]="isExpanded" [hideToggle]="isHideToggle">
    <ng-container headerTitle>
      <ng-content select="[panel-header-title]"></ng-content>
    </ng-container>

    <span panelDescription>
      <ng-content select="[accordion-panel-description]"></ng-content>
    </span>

    <ng-container headerIcon>
      <ng-content select="[panel-header-icon]"></ng-content>
    </ng-container>
  </app-accordion-panel-header>
</div>

<div role="region" class="accordion_panel__body" *ngIf="isExpanded" @body>
  <ng-content select="accordion-panel-content"></ng-content>
</div>
