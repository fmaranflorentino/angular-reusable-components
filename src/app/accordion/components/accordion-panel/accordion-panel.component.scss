.accordion_panel {
  box-shadow: 0 3px 1px -2px rgba(0, 0, 0, 0.2), 0 2px 2px 0 rgba(0, 0, 0, 0.14),
    0 1px 5px 0 rgba(0, 0, 0, 0.12);
  transition: height 225ms ease-out;

  cursor: pointer;

  display: block;

  &:first-of-type {
    border-top-right-radius: 4px;
    border-top-left-radius: 4px;
  }

  &:last-of-type {
    border-bottom-right-radius: 4px;
    border-bottom-left-radius: 4px;
  }

  &:not(:first-of-type) {
    margin-top: 10px;
  }


  .toggle {
    transition: all 0.3s ease-in;
  }

  &--expanded {



    .icon_content .toggle {
      transform: rotate(-180deg);
      transition: all 0.3s ease-in;
    }

    .accordion_panel__header_content {
      height: 64px;
      transition: height 225ms cubic-bezier(0.4, 0, 0.2, 1);
    }
  }

  &__header_content {
    padding: 0 24px;
    height: 45px;
    display: flex;
    align-items: center;
    justify-content: space-between;

    &--disabled {
      cursor: not-allowed;
      color: #999;
    }

    img {
      width: 14px;
    }

    .custom_icon {
      margin-right: 10px;
    }
  }

  &__body {
    padding: 0 24px 16px;
  }
}
