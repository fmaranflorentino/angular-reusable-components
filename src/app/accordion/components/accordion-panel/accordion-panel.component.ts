import { Component, Input, HostBinding, ViewEncapsulation } from '@angular/core';
import { AccordionBodyAnimation } from '../../accordion.animations';

@Component({
  selector: 'accordion-panel',
  templateUrl: './accordion-panel.component.html',
  styleUrls: ['./accordion-panel.component.scss'],
  encapsulation: ViewEncapsulation.None,
  animations: [ AccordionBodyAnimation ],
})
export class AccordionPanelComponent {
  @HostBinding('class.accordion_panel')
  accordion_panel = true;

  @HostBinding('class.accordion_panel--expanded')
  get expandedValue() {
    return this.isExpanded === true;
  };

  @HostBinding('class.accordion_panel__header__container--disabled')
  get disabledValue() {
    return this.isDisabled === true;
  }

  @Input()
  set expanded(value) {
    this._expanded = value;
  }

  @Input()
  set disabled(value: boolean) {
    this._disabled = value !== null && `${value}` !== 'false';
  }

  @Input()
  set hideToggle(value: boolean) {
    this._hideToggle = value !== null && `${value}` !== 'false';
  }

  private _expanded: boolean;

  private _disabled: boolean;

  private _hideToggle: boolean = false;


  get isExpanded(): boolean {
    return this._expanded;
  }

  get isDisabled(): boolean {
    return this._disabled;
  }

  get isHideToggle(): boolean {
    return this._hideToggle;
  }

  handlePanelInteraction(): void {
    if (this.isDisabled) {
      return;
    }

    this._expanded = !this._expanded;
  }

}
