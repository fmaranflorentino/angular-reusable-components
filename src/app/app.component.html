<section>
  <button customButton color="primary">
    primary
  </button>

  <button customButton color="secondary">
    secondary
  </button>

  <button customButton outline>
    outline
  </button>

  <button customButton outline>
    outline icon left
    <img src="/assets/send.svg" alt="" icon />
  </button>

  <button customButton color="primary" iconPosition="right">
    icon right
    <img src="/assets/send.svg" alt="" icon />
  </button>

  <button customButton disabled>
    disabled
  </button>
</section>
<hr />
<accordion>
  <accordion-panel>
    <h3 panel-header-title>Angular - Developer Experience</h3>
    <span accordion-panel-description
      >A meetup event for defining the future of developer experience.</span
    >
    <span class="material-icons" panel-header-icon>
      account_circle
    </span>

    <accordion-panel-content>content man</accordion-panel-content>
  </accordion-panel>

  <accordion-panel>
    <h3 panel-header-title>Angular - Developer Experience</h3>
    <span accordion-panel-description
      >A meetup event for defining the future of developer experience.</span
    >
    <span class="material-icons" panel-header-icon>
      account_circle
    </span>

    <accordion-panel-content>content man</accordion-panel-content>
  </accordion-panel>
</accordion>
