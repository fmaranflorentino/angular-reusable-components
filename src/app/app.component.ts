import { Component } from '@angular/core';
// import { testeServ } from './accordion/components/bottom-sheet/teste';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent {
  // constructor(private teste: testeServ) {
  //   this.teste.open();
  // }
}
