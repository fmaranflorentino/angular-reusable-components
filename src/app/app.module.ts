import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { NgModule } from '@angular/core';

import { AppComponent } from './app.component';
import { ButtonComponent } from './button/button.component';
import { AccordionComponent } from './accordion/accordion.component';
import { AccordionPanelComponent } from './accordion/components/accordion-panel/accordion-panel.component';
import { AccordionPanelHeaderComponent } from './accordion/components/accordion-panel-header/accordion-panel-header.component';
import { AccordionPanelDescriptionComponent } from './accordion/components/accordion-panel-description/accordion-panel-description.component';
import { CommonModule } from '@angular/common';
import { AccordionPanelContentComponent } from './accordion/components/accordion-panel-content/accordion-panel-content.component';
import { BottomSheatComponent } from './accordion/components/bottom-sheat/bottom-sheat.component';
import { BottomSheetComponent } from './accordion/components/bottom-sheet/bottom-sheet.component';
// import { testeServ } from './accordion/components/bottom-sheet/teste';
// import { OVERLAY_PROVIDERS } from '@angular/cdk/overlay';
import { PanelHeaderTitleComponent } from './accordion/components/accordion-panel-header/components/panel-header-title/panel-header-title.component';

@NgModule({
  declarations: [
    AppComponent,
    ButtonComponent,
    AccordionComponent,
    AccordionPanelComponent,
    AccordionPanelHeaderComponent,
    AccordionPanelDescriptionComponent,
    AccordionPanelContentComponent,
    BottomSheatComponent,
    BottomSheetComponent,
    PanelHeaderTitleComponent,
  ],
  imports: [BrowserModule, BrowserAnimationsModule, CommonModule],
  providers: [
    // testeServ,
    // OVERLAY_PROVIDERS,
  ],
  bootstrap: [AppComponent],
  exports: [
    ButtonComponent,
    AccordionComponent,
    AccordionPanelComponent,
    AccordionPanelHeaderComponent,
    AccordionPanelDescriptionComponent,
    AccordionPanelContentComponent,
    BottomSheatComponent,
    BottomSheetComponent,
    PanelHeaderTitleComponent,
  ],
  entryComponents: [BottomSheetComponent],
})
export class AppModule {}
