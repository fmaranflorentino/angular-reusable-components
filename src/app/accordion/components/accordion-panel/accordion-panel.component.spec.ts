import { SpectatorHost, createHostFactory } from '@ngneat/spectator/jest';
import { axe } from 'jest-axe';

import { AppModule } from '../../../app.module';
import { AccordionPanelComponent } from './accordion-panel.component';
import { APP_BASE_HREF } from '@angular/common';

const cases = {
  simple: `<accordion-panel></accordion-panel>`,
  expanded: `<accordion-panel [expanded]="true"></accordion-panel>`,
  disabled: `<accordion-panel disabled></accordion-panel>`,
  hideToggle: `<accordion-panel hideToggle></accordion-panel>`,
  complex: `
  <accordion-panel>
    <h3 panel-header-title>Angular - Developer Experience</h3>
    <span accordion-panel-description
      >A meetup event for defining the future of developer experience.</span
    >
    <span class="material-icons" panel-header-icon>
      account_circle
    </span>

    <accordion-panel-content>content man</accordion-panel-content>
  </accordion-panel>
  `
};

describe('AccordionPanelComponent', () => {
  let subject: SpectatorHost<AccordionPanelComponent>;
  const createHost = createHostFactory({
    component: AccordionPanelComponent,
    declareComponent: false,
    imports: [ AppModule ],
    providers: [{provide: APP_BASE_HREF, useValue: '/'}]
  });

  it('should create instance', () => {
    subject = createHost(cases.simple);
    expect(subject.component).toBeTruthy();
  });

  it('should create and match raw component snapshot', () => {
    subject = createHost(cases.simple);
    expect(subject.hostFixture).toMatchSnapshot();
  });

  it('should have class accordion_panel', () => {
    subject = createHost(cases.simple);
    expect(subject.hostFixture).toMatchSnapshot();
  });

  it('should have class accordion_panel--expanded', () => {
    subject = createHost(cases.expanded);
    expect(subject.hostFixture).toMatchSnapshot();
  });

  it('should have class accordion_panel__header__container--disabled', () => {
    subject = createHost(cases.disabled);
    expect(subject.hostFixture).toMatchSnapshot();
  });

  it('should render the component with all children', () => {
    subject = createHost(cases.complex);
    expect(subject.hostFixture).toMatchSnapshot();
  });

  it('should expand the panel', () => {
    subject = createHost(cases.simple);
    subject.component.handlePanelInteraction();
    subject.detectChanges();

    expect(subject.component.isExpanded).toBeTruthy();
  });

  it('should open and close the panel', () => {
    subject = createHost(cases.simple);
    subject.component.handlePanelInteraction();
    subject.detectChanges();

    expect(subject.component.isExpanded).toBeTruthy();

    subject.component.handlePanelInteraction();
    subject.detectChanges();

    expect(subject.component.isExpanded).toBeFalsy();
  });

  it('should add true to disable property', () => {
    subject = createHost(cases.disabled);
    expect(subject.component.isDisabled).toBeTruthy();
  });

  it('should add true to expand property', () => {
    subject = createHost(cases.expanded);
    expect(subject.component.isExpanded).toBeTruthy();
  });

  it('should add true to hideToggle property', () => {
    subject = createHost(cases.hideToggle);
    expect(subject.component.isHideToggle).toBeTruthy();
  });

  it('should verify component accessibility', async () => {
    subject = createHost(cases.simple);
    expect(await axe(subject.hostFixture.nativeElement)).toHaveNoViolations();
  });
});
