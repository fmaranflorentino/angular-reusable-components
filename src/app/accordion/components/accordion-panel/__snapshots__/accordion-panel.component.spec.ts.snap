// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionPanelComponent should create and match raw component snapshot 1`] = `
<ng-component>
  <accordion-panel
    class="accordion_panel"
  >
    <div
      class="accordion_panel__header__container"
    >
      <app-accordion-panel-header
        class="accordion_panel__header"
      >
        <div
          class="accordion_panel__header__wrapper"
        >
          
          <span
            paneldescription=""
          />
          <div
            class="custom_icon__container"
          >
            <span
              class="custom_icon"
            >
              
            </span>
            
            <span
              class="material-icons toggle"
            >
               keyboard_arrow_down 
            </span>
          </div>
        </div>
      </app-accordion-panel-header>
    </div>
    
  </accordion-panel>
</ng-component>
`;

exports[`AccordionPanelComponent should have class accordion_panel 1`] = `
<ng-component>
  <accordion-panel
    class="accordion_panel"
  >
    <div
      class="accordion_panel__header__container"
    >
      <app-accordion-panel-header
        class="accordion_panel__header"
      >
        <div
          class="accordion_panel__header__wrapper"
        >
          
          <span
            paneldescription=""
          />
          <div
            class="custom_icon__container"
          >
            <span
              class="custom_icon"
            >
              
            </span>
            
            <span
              class="material-icons toggle"
            >
               keyboard_arrow_down 
            </span>
          </div>
        </div>
      </app-accordion-panel-header>
    </div>
    
  </accordion-panel>
</ng-component>
`;

exports[`AccordionPanelComponent should have class accordion_panel__header__container--disabled 1`] = `
<ng-component>
  <accordion-panel
    class="accordion_panel accordion_panel__header__container--disabled"
    disabled=""
  >
    <div
      class="accordion_panel__header__container"
    >
      <app-accordion-panel-header
        class="accordion_panel__header"
      >
        <div
          class="accordion_panel__header__wrapper"
        >
          
          <span
            paneldescription=""
          />
          <div
            class="custom_icon__container"
          >
            <span
              class="custom_icon"
            >
              
            </span>
            
            <span
              class="material-icons toggle"
            >
               keyboard_arrow_down 
            </span>
          </div>
        </div>
      </app-accordion-panel-header>
    </div>
    
  </accordion-panel>
</ng-component>
`;

exports[`AccordionPanelComponent should have class accordion_panel--expanded 1`] = `
<ng-component>
  <accordion-panel
    class="accordion_panel accordion_panel--expanded"
  >
    <div
      class="accordion_panel__header__container"
    >
      <app-accordion-panel-header
        class="accordion_panel__header accordion_panel__header--toggleanime"
      >
        <div
          class="accordion_panel__header__wrapper"
        >
          
          <span
            paneldescription=""
          />
          <div
            class="custom_icon__container"
          >
            <span
              class="custom_icon"
            >
              
            </span>
            
            <span
              class="material-icons toggle"
            >
               keyboard_arrow_down 
            </span>
          </div>
        </div>
      </app-accordion-panel-header>
    </div>
    
    <div
      class="accordion_panel__body ng-trigger ng-trigger-body ng-animating"
      role="region"
      style="animation: 500ms ease-in 0ms 1 normal forwards gen_css_kf_0; animation-play-state: running;"
    />
  </accordion-panel>
</ng-component>
`;

exports[`AccordionPanelComponent should render the component with all children 1`] = `
<ng-component>
  <accordion-panel
    class="accordion_panel"
  >
    <div
      class="accordion_panel__header__container"
    >
      <app-accordion-panel-header
        class="accordion_panel__header"
      >
        <div
          class="accordion_panel__header__wrapper"
        >
          
          <h3
            class="accordion_panel__header__title"
            panel-header-title=""
          >
            Angular - Developer Experience
          </h3>
          <span
            paneldescription=""
          >
            <span
              accordion-panel-description=""
              class="accordion_panel__header__description"
            >
              A meetup event for defining the future of developer experience.
            </span>
          </span>
          <div
            class="custom_icon__container"
          >
            <span
              class="custom_icon"
            >
              
              <span
                class="material-icons"
                panel-header-icon=""
              >
                 account_circle 
              </span>
            </span>
            
            <span
              class="material-icons toggle"
            >
               keyboard_arrow_down 
            </span>
          </div>
        </div>
      </app-accordion-panel-header>
    </div>
    
  </accordion-panel>
</ng-component>
`;
